### ISSUE TITLE
TypeError: Cannot convert undefined or null to object (When I try run it on web browser with expo)

### ISSUE THREAD

**Issue by codingki** [nandorojo/moti#14]
![image](https://user-images.githubusercontent.com/39829726/107361752-72838900-6b0a-11eb-8e70-1f618bd6a521.png)

**Comment by nandorojo** [comment]
Duplicate of https://github.com/nandorojo/moti/issues/10

If you want help, please upload a minimal repo example.

This should be solved in the latest version of Moti, is it not?

**Comment by nandorojo** [comment]
This seems like a reanimated issue; the stack trace is showing `useAnimatedStyle`, which is a Reanimated hook.

**Comment by codingki** [comment]
> Duplicate of #10
> 
> If you want help, please upload a minimal repo example.
> 
> This should be solved in the latest version of Moti, is it not?
here is the repo, I try to load minimal example of moti but still not working on the web
[https://github.com/codingki/react-native-touchable-moti-example](https://github.com/codingki/react-native-touchable-moti-example)

**Comment by codingki** [comment]
I use the latest version of Moti, its 0.4.1 
here is the repo, I try to load minimal example of Moti but still not working on the web
https://github.com/codingki/react-native-touchable-moti-example

**Comment by codingki** [comment]
> This seems like a reanimated issue; the stack trace is showing `useAnimatedStyle`, which is a Reanimated hook.

Yeah I think so

**Comment by nandorojo** [comment]
I'll see if I can get to the bottom of it, thanks!

**Comment by nandorojo** [comment]
@cmaycumber interestingly, this problem happens in an expo app, but not in our `with-expo` folder. Any reason come to mind that our example might not have the same problem? Is there any magic in the webpack?

I'm going to try upgrading RNW to see if that helps.

**Comment by nandorojo** [comment]
To reproduce this bug:

`npx create-react-native-app -t with-moti`
`yarn web`

**Comment by cmaycumber** [comment]
> @cmaycumber interestingly, this problem happens in an expo app, but not in our `with-expo` folder. Any reason come to mind that our example might not have the same problem? Is there any magic in the webpack?

Maybe it could have something to do with the files after they're built by react-native-bob. The webpack resolves directly to the src/index.ts. That would be my best guess without looking at it.  

I'll try to reproduce this on my end real quick. Let me know if upgrading RNW helps.

**Comment by cmaycumber** [comment]
What if we added the babel plugin for react-native-reanimated into the babel.config.js at the root that bob uses for the build?

**Comment by nandorojo** [comment]
Ohh interesting idea. Does Babel get used at a build step?

**Comment by nandorojo** [comment]
I find it odd that Moti doesn't have this issue in my Expo + Next.js app.

**Comment by cmaycumber** [comment]
> I find it odd that Moti doesn't have this issue in my Expo + Next.js app.

Yeah, this is odd. I thought the same thing when you mentioned you use it in your app. I haven't started using it in mine but I expect I'd see the same problem. 

> Ohh interesting idea. Does Babel get used at a build step?

I think bob uses babel to compile all the source files when you run `bob build`. I could be mistaken though. I'm going to try to play around with the babel config that bob uses a bit.

**Comment by nandorojo** [comment]
It looks like Expo Web uses the lib/module folder. I use next-transpile-modules, so my guess is that next is using my Babel config on Moti and that's why it works.

@cmaycumber as you suggested, I think adding the Babel config in Bob could work. Let me know if you confirm that they use it.

**Comment by cmaycumber** [comment]
This is from the bob docs for the module target:

> Enable compiling source files with Babel and use ES module system. This is essentially same as the commonjs target and accepts the same options, but leaves the import/export statements in your code.
> 
> This is useful for bundlers which understand ES modules and can tree-shake. The output file should be referenced in the module field of package.json.

Sadly, changing the babel config hasn't worked for me yet. After packing and running it locally but I'm going to clear my `node_modules` just in case.

**Comment by nandorojo** [comment]
Looking into the lib code of modules, it seems like it's no longer calling `useAnimatedStyle`, but instead a weird looking output from Bob's Babel build. My guess is, Reanimated's Babel plugin might be skipping over this because it doesn't recognize it.

**Comment by nandorojo** [comment]
Alright, well I got it to (kind of) work by adding this to the example:

1. `yarn add -D babel-loader`
2. Create a custom `webpack.config.js` file:

`webpack.config.js`

```js
const createExpoWebpackConfigAsync = require("@expo/webpack-config");

module.exports = async function (env, argv) {
  const config = await createExpoWebpackConfigAsync(env, argv);

  config.module.rules.push({
    test: /\.(js|ts|tsx)$/,
    exclude: /node_modules\/(?!(moti|@motify)\/).*/,
    use: "babel-loader",
  });

  return config;
};
```

I don't love this solution, since it adds quite a bit of config to the user, and I'm not knowledgeable enough about webpack/babel to know if this is a good idea. I have a feeling this is not the right way to use the babel loader, since I believe it's applying to all files except the moti ones. Honestly, not sure.

That said, it does make the example start, so there is some progress.

Basically, the issue is, Reanimated's babel plugin is not applying to `@motify/core` (or `moti` for that matter.) We need it to apply to `@motify/core/lib/module`. Maybe there's a simpler way to do that than editing webpack?

**Comment by nandorojo** [comment]
I opened a discussion here: https://github.com/software-mansion/react-native-reanimated/discussions/1715

**Comment by nandorojo** [comment]
@simpleshadow I saw you liked my comment – did this solution work for you, by chance?

**Comment by nandorojo** [comment]
I think I found the right answer, thanks to the `@expo/webpack-config` [docs](https://github.com/expo/expo-cli/tree/master/packages/webpack-config#expobabeloptions):

```js
const createExpoWebpackConfigAsync = require("@expo/webpack-config");

module.exports = async function (env, argv) {
  const config = await createExpoWebpackConfigAsync(
    {
      ...env,
      babel: { dangerouslyAddModulePathsToTranspile: ["moti"] },
    },
    argv
  );

  return config;
};
```

Expo lets you pass a custom set of node modules to transpile with `dangerouslyAddModulePathsToTranspile`, so that did it!

Could everyone here please try this solution?

**Comment by nandorojo** [comment]
Solution added to docs: https://moti-5qbvkhyh8.vercel.app/web#expo-web-support

![Screen Shot 2021-02-15 at 2 35 31 PM](https://user-images.githubusercontent.com/13172299/107986918-13ef5c80-6f9b-11eb-8818-2df763f4716c.png)

**Comment by codinsonn** [comment]
Any pointers on avoiding `useAnimatedStyle()` breaking due to `updater._closure` being `null` in Next.js (12) serverside envs?

I'm not entirely sure where that `._closure` is even supposed to come from, but my stack trace tells me the `updater` is something that's passed from @motify/core/lib/commonjs/use-map-animate-to-style.js

<details>
<summary>Show stack trace</summary>

<img width="1007" alt="Screenshot 2021-10-30 at 16 31 26" src="https://user-images.githubusercontent.com/5967956/139537363-cf816869-7182-487f-9eee-1acd9443d05c.png">

</details>

To be clear, it does work on Expo Go & Expo Web (SDK 43) now, thanks to this GH issue, but not yet for Next.js

If no suggestion comes to mind, should I make another issue as this is related to use of moti in next.js?
(In which case I'll see if I can try to extract a minimal repro from my project)

## Things already done / tried:

- All instructions from: https://moti.fyi/web
- All instructions from: https://moti.fyi/next
- All things suggested in this thread
- Updating `moti`, `react-native-reanimated` & `@expo/next-adapter` with `expo install ...` (no dice)
- Updating `moti`, `react-native-reanimated` & `@expo/next-adapter` with `npm install ...@latest`
- Making sure `'react-native-reanimated/plugin'` is added to plugins in `babel.config.js` (already the case)

## Used configs

<details>
<summary>babel.config.js</summary>

```js
module.exports = {
    presets: ['@expo/next-adapter/babel'],
    plugins: [
        'react-native-reanimated/plugin',
        '@babel/plugin-proposal-class-properties',
    ],
};
```

</details>

<details>
<summary>next.config.js</summary>

```js
onst { withExpo } = require('@expo/next-adapter');
const withFonts = require('next-fonts');
const withPlugins = require('next-compose-plugins');

const withTM = require('next-transpile-modules')([
  'moti',
  '@motify/core',
  '@motify/components',
]);

module.exports = withPlugins(
  [withTM, withFonts, [withExpo, { projectRoot: __dirname }]],
  {
    typescript: {
      ignoreBuildErrors: true,
    },
  },
);
```

</details>

**Comment by nandorojo** [comment]
@codinsonn any shot you could try to clone this and use it to repro? https://github.com/axeldelafosse/expo-next-monorepo-example

**Comment by codinsonn** [comment]
> @codinsonn any shot you could try to clone this and use it to repro? https://github.com/axeldelafosse/expo-next-monorepo-example

On it ⏳

**Comment by nandorojo** [comment]
I actually had this issue myself yesterday, but with normal reanimated. It seemed that passing `null` to `useDerivedValue` was breaking it for some reason. When I changed `null` to a number outside of the worklet, it got fixed.

**Comment by codinsonn** [comment]
@nandorojo Hmm, I seem to be unable to reproduce the issue with the monorepo you suggested 🤷‍♂️
https://github.com/codinsonn/moti-next12-monorepro

But, since it does seem to work there, I guess I'll try copying some of the configs and versions used to see if it fixes it in my current project. If I find what eventually was the issue, I'll post an update here ⏳

**Comment by nandorojo** [comment]
Yeah, I think it’s because expo and next requires a monorepo with separate entry points and very particular Babel configs (root of the monorepo, and a unique config in each entry point)
